### PR TITLE
Add configuration option to suppress require cycle warnings

### LINKED ISSUE #287: Require cycle warnings should be opt-in when the source is node_modules
**Do you want to request a *feature* or report a *bug*?**

Feature

**What is the current behavior?**

Lots of warnings about require cycles from libraries in node_modules in the react-native ecosystem, coming from the following code: https://github.com/facebook/metro/blob/dc8d85911ff6bf6487b7e3aef6b004df4bdcf17f/packages/metro/src/lib/polyfills/require.js#L158-L174

**What is the expected behavior?**

Don't warn users about require cycles in libraries in node_modules unless the user opts-in. Warnings about code that you don't own, especially when the warnings are non-essential, create unnecessary noise.

**Please provide your exact Metro configuration and mention your Metro, node, yarn/npm version and operating system.**

React Native 0.57 default configuration and Metro version (0.47.1). Other information irrelevant.

### PR DESCRIPTION
**Pull request by hsource**
**Summary**

In many projects there are require cycles in the node_modules dependencies that don't really need to be fixed and that don't seem to affect the functioning of the app. This change adds an option to let users selectively suppress them.

Unlike many of the workarounds in #287, this requires opting in per-package, so that you're never blindsided by bugs caused by require cycles in subpackages you own or maintain.

Closes #287

**Test plan**

1. Built metro in my own branch using `yarn build && yarn lerna run prepare-release`
2. Copied over changed files (require.js, getPreludeCode.js, getPrependedScripts.js) into my own node_modules
3. Changed metro.config.js and ran `react-native start` with the extra config below

```js
module.exports = {
  // ...
  resolver: {
    // ...
    ignoreRequireCyclePrefixes: [
      'node_modules/react-native-firebase',
      'node_modules/react-native-maps',
      'node_modules/rn-fetch-blob',
      'node_modules/quill-delta',
    ],
  },
};
```

### Before

```
 WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/Blob.js -> node_modules/rn-fetch-blob/index.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/XMLHttpRequest.js -> node_modules/rn-fetch-blob/index.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/Fetch.js -> node_modules/rn-fetch-blob/index.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/react-native-firebase/dist/utils/apps.js -> node_modules/react-native-firebase/dist/modules/core/app.js -> node_modules/react-native-firebase/dist/utils/apps.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/react-native-firebase/dist/modules/admob/index.js -> node_modules/react-native-firebase/dist/modules/admob/Interstitial.js -> node_modules/react-native-firebase/dist/modules/admob/index.js
```

### After

No warnings

### CODE REVIEW

File: packages/metro/src/lib/getPreludeCode.js [hunk 1]
```diff
@@ -14,17 +14,24 @@ function getPreludeCode({
...
+    // Ensure these variable names match the ones referenced in metro-runtime
+    // require.js
     '__BUNDLE_START_TIME__=this.nativePerformanceNow?nativePerformanceNow():Date.now()',
     `__DEV__=${String(isDev)}`,
     ...formatExtraVars(extraVars),
     'process=this.process||{}',
     `__METRO_GLOBAL_PREFIX__='${globalPrefix}'`,
+    `__IGNORE_REQUIRE_CYCLE_PREFIXES__=${JSON.stringify(
+      ignoreRequireCyclePrefixes,
+    )}`,
```
[the remark this answers is not in the document]
Comment (hsource) [reply packages/metro/src/lib/getPreludeCode.js]: Done in the latest change!

**Comment by hsource** [comment]
> @hsource this is awesome, thanks for the PR! There are some details to discuss but I love the idea of making this warning less noisy and more actionable.
> 
> 1. I would go further and ignore require cycles in `node_modules` _by default_. It's a safe assumption that users don't have direct control over cycles in third-party modules, and that code that's already published to npm was at least minimally tested so the warning is a false positive there.
> 2. I wouldn't limit the filtering capability to prefixes. How about changing this to accept regexes? (We can call it e.g. `requireCycleIgnorePatterns`.) That way we can check for `node_modules` anywhere in the project, not just at the root.
> 3. Let's make sure this works out of the box for Windows paths. One way would be to use regexes as per (2), and make the default pattern check for both forward and backward slashes around `node_modules`.
> 4. I wonder if we should check all the modules in a cycle against the patterns (and suppress the warning if there are any matches?) or whether checking just the first module is enough. We can start with the current solution, though.

@motiz88 Thanks for the review and sorry for the delay. I had some spare time and just cleaned up this PR.

I applied all of your changes. The default RegExp looks a bit messy, but I've verified that it works with both slashes and backslashes. Can you take another look?

**Comment by Latropos** [comment]
@hsource Thank you for your work 💚, any update on it?

**Comment by hsource** [comment]
I think it's just awaiting a pull. @motiz88 or anyone else can take a look when they have time!

**Comment by Latropos** [comment]
@motiz88 🥺 🙏 💔❓
Could sb look at this PR?
@mroch @pieterv @robhogan

**Comment by robhogan** [comment]
Tests are failing because `acorn@^5.1.2`, which we use only as a dev dependency, is too old to support `for..of` syntax. I'll update that first.

**Comment by hsource** [comment]
> Am I ok to pull that change in to your PR here and re-import for review?

@robhogan Yes! Go for it. Thanks a ton for improving on this

**Pull request closed without merging**